packages/react-form-renderer/src/form-renderer/form-renderer.js

```diff
@@ -1,77 +1,132 @@
-import React, { useState, useRef } from 'react';
-import Form from '../form';
 import arrayMutators from 'final-form-arrays';
-import PropTypes from 'prop-types';
 import createFocusDecorator from 'final-form-focus';
+import PropTypes from 'prop-types';
+import React, {useCallback, useMemo, useRef, useState} from 'react';
 
+import defaultSchemaValidator from '../default-schema-validator';
+import defaultValidatorMapper from '../validator-mapper';
+import Form from '../form';
 import RendererContext from '../renderer-context';
 import renderForm from './render-form';
-import defaultSchemaValidator from '../default-schema-validator';
 import SchemaErrorComponent from './schema-error-component';
-import defaultValidatorMapper from '../validator-mapper';
+
+
+const isFunc = (fn) => (typeof fn === 'function');
 
 const FormRenderer = ({
+  actionMapper,
+  children,
+  clearedValue,
+  clearOnUnmount,
   componentMapper,
+  decorators,
   FormTemplate,
-  onSubmit,
+  FormTemplateProps,
+  mutators,
   onCancel,
+  onError,
   onReset,
-  clearOnUnmount,
-  subscription,
-  clearedValue,
+  onSubmit,
   schema,
-  validatorMapper,
-  actionMapper,
   schemaValidatorMapper,
+  subscription,
+  validatorMapper,
   ...props
 }) => {
   const [fileInputs, setFileInputs] = useState([]);
   const registeredFields = useRef({});
   const focusDecorator = useRef(createFocusDecorator());
-  let schemaError;
+  const validatorMapperMerged = useMemo(() => ({
+    ...defaultValidatorMapper,
+    ...validatorMapper
+  }), [validatorMapper]);
+  const mutatorsMerged = useMemo(() => ({
+    ...arrayMutators,
+    ...mutators
+  }), [mutators]);
+  const decoratorsMerged = useMemo(() => ([
+    focusDecorator.current,
+    ...(Array.isArray(decorators) ? decorators : [])
+  ]), [decorators]);
+
+  const handleSubmitCallback = useCallback((values, formApi, ...args) => {
+    return !isFunc(onSubmit) ? undefined : onSubmit(values, {...formApi, fileInputs}, ...args);
+  }, [onSubmit, fileInputs]);
+
+  const handleCancelCallback = useCallback((getState) => (...args) => {
```

Comment:
@zgover This causes the issue with the unexpected `Cancel` button in FormTemplates.

The common `FormTemplate` is checking if the function exists and if it does, it will render the `cancel` button. As `useCallback` always returns a function regardless of the callback, it confuses the templates.

Reply:
@Hyperkid123 I'll leave that to you to propose a solution. It appears there are a few areas that need addressing with the test cancel validation. Personally, I think this should only be validating the `onCancel` pre and post logic, if currently the default behavior displays a cancel element, that should instead be decided by the lib consumer.

Reply:
Yes, they can decide. We are talking about common default templates that are not mandatory to use. So ultimately the decision is on the consumer. I would not like to introduce an entirely new prop that says `enableCancel` or something similar.

You should be able to fix with with a simple `onCancel: onCancel && handleCancelCallback(getState),` assignment